kotlin/workflow-runtime/src/main/java/com/squareup/workflow/WorkflowHost.kt

```diff
@@ -36,27 +45,63 @@ import kotlin.coroutines.coroutineContext
 private val DEFAULT_WORKFLOW_COROUTINE_NAME = CoroutineName("WorkflowHost")
 
 /**
- * Provides a stream of [updates][Update] from a tree of [Workflow]s.
+ * Provides streams of [renderings and snapshots][RenderingAndSnapshot] and outputs from a tree
+ * of [Workflow]s.
  *
  * Create these by injecting a [Factory] and calling [run][Factory.run].
+ *
+ * [start] must be called to start the workflow running, and then [cancel] must be called to stop
+ * it.
  */
 interface WorkflowHost<out OutputT : Any, out RenderingT> {
 
   /**
-   * Output from a [WorkflowHost]. Emitted from [WorkflowHost.updates] after every render pass.
+   * Emitted from [renderingsAndSnapshots] after every render pass.
    */
-  data class Update<out OutputT : Any, out RenderingT>(
+  data class RenderingAndSnapshot<out RenderingT>(
     val rendering: RenderingT,
-    val snapshot: Snapshot,
-    val output: OutputT? = null
+    val snapshot: Snapshot
   )
 
   /**
-   * Stream of [updates][Update] from the root workflow.
+   * Stream of [renderings and snapshots][RenderingAndSnapshot] from the root workflow.
+   * Renderings and snapshots are always taken together, so they are emitted together.
+   *
+   * Once the [WorkflowHost] is [started][start], this Flow will immediately emit the first
+   * rendering to any collectors, and then emit new renderings/snapshots any time something happens
+   * within the workflow tree that causes a new render pass.
    *
-   * This is *not* a broadcast channel, so it should only be read by a single consumer.
+   * New collectors arriving after calling [start] will always immediately get the last rendering
+   * and snapshot.
+   *
+   * If any workflow or worker throws an exception, it will be re-thrown to collectors of this Flow
+   * (although it may be wrapped in one or more [CancellationException]s).
    */
-  val updates: ReceiveChannel<Update<OutputT, RenderingT>>
+  @UseExperimental(ExperimentalCoroutinesApi::class)
+  val renderingsAndSnapshots: Flow<RenderingAndSnapshot<RenderingT>>
+
+  /**
+   * Stream of outputs from the root workflow.
+   *
+   * This Flow is hot – it does *not* replay any old outputs to new collectors.
+   *
+   * If any workflow or worker throws an exception, it will be re-thrown to collectors of this Flow
+   * (although it may be wrapped in one or more [CancellationException]s).
+   */
+  @UseExperimental(ExperimentalCoroutinesApi::class)
+  val outputs: Flow<OutputT>
+
+  /**
+   * Start the workflow. [renderingsAndSnapshots] and [outputs] won't emit anything until this
+   * is called.
+   */
+  fun start()
+
+  /**
+   * Stop the workflow tree, cancelling every worker and workflow. This will cause the
+   * [renderingsAndSnapshots] and [outputs] Flows to complete.
+   */
+  fun cancel()
```

Comment:
An alternative design to having an explicit cancel method is to make `start` suspending and call it `run` or something. It suspends until the workflow throws an exception or the coroutine is cancelled. Feels a bit safer, since if you use structured concurrency you can't forget to cancel the runtime.

We could also then use the context from which `run` is called to determine the context for the session, although I'm not sure I like moving the context out of the factory – that's the whole reason the factory exists.

Reply:
I don't really have the instincts for this yet, but this kind of sounds like an all or nothing situation: if we're not comfortable making `suspend fun run()` work in the calling context, it seems wrong to make it `suspend`. My gut is to declare victory and merge this.

Reply:
Yea agree on that first point.

I'm still going to change this API though. `start()` will return a `Job`, and I'll delete `cancel()`. Then up in `WorkflowRunner`, I'll give it its own start method that will hang onto the `Job` and cancel it when the view model is `onCleared`. `WorkflowRunner` needs a start method too so that consumers can subscribe to `output` before starting a workflow that could potentially emit an output right away.